src/bitcoin-util.cpp

```diff
@@ -0,0 +1,222 @@
+        tfm::format(std::cerr, "Error: %s\n", e.what());
+        return EXIT_FAILURE;
+    }
+
+    if (argc < 2 || HelpRequested(gArgs)) {
+        // First part of help message is specific to this utility
+        std::string strUsage = PACKAGE_NAME " bitcoin-tx utility version " + FormatFullVersion() + "\n\n" +
+            "Usage:  bitcoin-util [options] [commands]  Do stuff\n" +
+            "\n";
+        strUsage += gArgs.GetHelpMessage();
+
+        tfm::format(std::cout, "%s", strUsage);
+
+        if (argc < 2) {
+            tfm::format(std::cerr, "Error: too few parameters\n");
+            return EXIT_FAILURE;
+        }
+        return EXIT_SUCCESS;
+    }
+    return CONTINUE_EXECUTION;
+}
+
+static void grind_task(uint32_t nBits, CBlockHeader& header_orig, uint32_t offset, uint32_t step, std::atomic<bool>& found)
+{
+    arith_uint256 target;
+    bool neg, over;
+    target.SetCompact(nBits, &neg, &over);
+    if (target == 0 || neg || over) return;
+    CBlockHeader header = header_orig; // working copy
+    header.nNonce = offset;
+
+    uint32_t finish = std::numeric_limits<uint32_t>::max() - step;
+    finish = finish - (finish % step) + offset;
+
+    while (!found && header.nNonce < finish) {
+        const uint32_t next = (finish - header.nNonce < 5000*step) ? finish : header.nNonce + 5000*step;
+        do {
+            if (!(UintToArith256(header.GetHash()) > target)) {
+                if (!found.exchange(true)) {
+                    header_orig.nNonce = header.nNonce;
+                }
+                return;
+            }
+            header.nNonce += step;
+        } while(header.nNonce != next);
+    }
+}
+
+static int Grind(int argc, char* argv[], std::string& strPrint)
+{
+    if (argc != 1) {
+        strPrint = "Must specify block header to grind";
+        return 1;
+    }
+
+    CBlockHeader header;
+    if (!DecodeHexBlockHeader(header, argv[0])) {
+        strPrint = "Could not decode block header";
+        return 1;
+    }
+
+    uint32_t nBits = header.nBits;
+    std::atomic<bool> found{false};
+
+    std::vector<std::thread> threads;
+    int n_tasks = std::max(1u, std::thread::hardware_concurrency());
+    for (int i = 0; i < n_tasks; ++i) {
+        threads.emplace_back( grind_task, nBits, std::ref(header), i, n_tasks, std::ref(found) );
+    }
+    for (auto& t : threads) {
+        t.join();
+    }
+    if (!found) {
+        strPrint = "Could not satisfy difficulty target";
+        return 1;
+    }
+
+    CDataStream ss(SER_NETWORK, PROTOCOL_VERSION);
+    ss << header;
+    strPrint = HexStr(ss);
+    return 0;
+}
+
+static int CommandLineUtil(int argc, char* argv[])
+{
+    if (argc <= 1) return 1;
+
+    std::string strPrint;
+    int nRet = 0;
+
+    try {
+        while (argc > 1 && IsSwitchChar(argv[1][0]) && (argv[1][1] != 0)) {
+            --argc;
+            ++argv;
+        }
+
+        char* command = argv[1];
+        if (strcmp(command, "grind") == 0) {
```

Comment:
Any reason to not use our ArgsManager here, so that `-help` documentation will be created?

Reply:
I don't think ArgsManager supports subcommands that have different suboptions?

Reply:
It would support `-grind=block_header_hex`. Is something else needed here that I am missing?

Reply:
I was thinking it would make sense for `bitcoin-util` would have subcommands like git, each with their own different options (like `--amend` for `git commit` vs `--continue` for `git rebase`). `bitcoin-util grind` could have a `-jobs=8` argument specifying how many jobs to run in parallel, eg, which presumably wouldn't be appropriate elsewhere.

Reply:
The wallet tool is doing exactly that already with argsman. E.g. it has the `info` subcommand, or a dump subcommand with conditional options like -dumpfile (#19137)

Reply:
This seems worth addressing still: https://github.com/bitcoin/bitcoin/pull/19937/commits/e6473d274fe01d58c9eed75bc56415fcce6d63db#r492742101